types/prismjs/index.d.ts

```diff
@@ -1,228 +1,435 @@
+ *
+ * This method could be useful in other contexts as well, as a very crude parser.
  *
  * @param text A string with the code to be highlighted.
- * @param grammar An object containing the tokens to use. Usually a language definition like
- * Prism.languages.markup
- * @returns An array of strings, tokens (class Prism.Token) and other arrays.
+ * @param grammar An object containing the tokens to use.
+ *
+ * Usually a language definition like `Prism.languages.markup`.
+ * @returns An array of strings, tokens and other arrays.
  */
-export function tokenize(text: string, grammar: LanguageDefinition): Array<Token | string>;
-
-export function fileHighlight(): void;
-
-export interface Environment {
-    element?: Element;
-    language?: LanguageDefinition;
-    grammar?: any;
-    code?: any;
-    highlightedCode?: any;
-    type?: string;
-    content?: string;
-    tag?: string;
-    classes?: string[];
-    attributes?: any;
-    parent?: Element;
+export function tokenize(
+	text: string,
+	grammar: Grammar
+): Array<string | Token>;
+
+export interface Environment extends Record<string, any> {
+	selector?: string;
+	element?: Element;
+	language?: string;
+	grammar?: Grammar;
+	code?: string;
+	highlightedCode?: string;
+	type?: string;
+	content?: string;
+	tag?: string;
+	classes?: string[];
+	attributes?: Record<string, string>;
+	parent?: Array<string | Token>;
 }
 
-export interface Identifier {
-    value: number;
+export namespace util {
+	interface Identifier {
+		value: number;
+	}
+
+	/** Encode raw strings in tokens in preparation to display as HTML */
+	function encode(tokens: TokenNode): TokenNode;
+
+	/** Determine the type of the object */
+	function type(o: null): "Null";
+	function type(o: undefined): "Undefined";
+	// tslint:disable:ban-types
+	function type(o: boolean | Boolean): "Boolean";
+	function type(o: number | Number): "Number";
+	function type(o: string | String): "String";
+	function type(o: Function): "Function";
+	// tslint:enable:ban-types
+	function type(o: RegExp): "RegExp";
+	function type(o: any[]): "Array";
+	function type(o: any): string;
+
+	/** Get the unique id of this object or give it one if it does not have one */
+	function objId(obj: any): Identifier;
+
+	/** Deep clone a language definition (e.g. to extend it) */
+	function clone<T>(o: T): T;
 }
 
-export interface Util {
-    /** Encode raw strings in tokens in preparation to display as HTML */
-    encode(tokens: TokenNode): TokenNode;
-
-    /** Determine the type of the object */
-    type(o: any): string;
-
-    /** Get the unique id of this object or give it one if it does not have one */
-    objId(obj: any): Identifier;
-
-    /** Deep clone a language definition (e.g. to extend it) */
-    clone(o: LanguageDefinition): LanguageDefinition;
+export type GrammarValue = RegExp | TokenObject | Array<RegExp | TokenObject>;
+export type Grammar = GrammarRest & Record<string, GrammarValue>;
```

Comment:
Doesn't this mean that the type of `Grammar#rest` will be `GrammarValue & Grammar` which is `never`?

I'm asking because I had a similar problem a while ago and I have yet to find a solution.

Reply:
No, this is different.

Merging two types using `&` when one is defined using an index signature will cause the definite one to always override the index signature.

Result: `Grammar#rest` has the type of `Grammar`.